README
Assignment 1.10/2- Stephanie Engelhardt
The Basilisk Game (aka a cooler version of the snake game):

For this assignment, I made a snake game that is Harry Potter themed. The game comes with
a make file, and can be ran using "./main". The background of the story is that you are the
basilisk within the chamber of secrets, and you have to go around and "petrify" each of the
different muggle borns by walking over them. To control the basilisk, you can use the
arrow keys or k (for up), j (for down), h(for left), or l (for right). If you don't move
within 1/4 of a second, the basilisk will automatically start moving- it moves diagnolly
way in order to make it harder. If you run over the eggs following you or if you go out of
bounds (by going onto the tiles marked "X", you die. Whenever you die or if you win, there
is ASCII art displayed. In addition, there is ASCII art at the beginning of the game with an
option to go to the instructions page. When you pick up a new egg, the program picks from a
random list of muggle born names from Harry Potter and tells you who you petrified on the 
right hand side of the chamber. That message stays there until the next time you petrify
someone. I used our professors code for the io_queue messages (which is commented within the
code that I used it above the functions I used). I also have colors within the project 
(green to represent the basilisk and blue for the title 'chamber of secrets'). Within my 
code I use something similar as the dungeon type in previous assignments for the map within
this project and then I have classes for the eggs and the snake. The muggles are placed 
randomly throughout the chamber, as well as the basilisk's starting position. 

CHANGELOG
Basilisk Game
By Stephanie Engelhardt

04/16/2017:
-Came up with the game idea
-Began to work out the different classes I would need: a map, eggs, and a snake
-Made the .cpp files for a map, a snake, and a main, put the egg class within the snake.h
file because it was very small and didn't need its own file

04/17/2017:
-Initialized the map and randomly put the muggle borns and the basilisk's starting position
within it
-Used ncurses to allow user input for to move the basilisk, it goes over the egg and removes 
it from the map, but doesn't make it follow the basilisk yet [the eggs are called "traps" 
before they are picked up]

04/18/2017:
-Made ASCII art for the beginning of the game as well as an instructions page that can
be viewed from the beginnning of the game. 
-Added an old position of x and y to the snake class in order to know where to go when
a user doesn't input anything-- I go diagonally based off of their movement, but I'm not
sure yet how to make it so the game doesn't stop to wait for the user input. 

04/19/2017:
-Found out about the select function in order to not wait for user input. I found this in the
io.cpp file from the dungeon project and used this to implement my own code. This allows my
automatic movement to begin working. 
-Got the eggs to start to follow the snake using the snake's position (not the previous 
position because I move the eggs first), and then move the other eggs within the vector to
the position of the egg before it in line. 

04/20/2017:
-Added a table of the different muggle borns to the map.cpp file
-Made ASCII art for the victory and losing pages

04/21/2017:
-Added colors to the Basilisk and following eggs as well as the title on the chamber
-Added queue messages using the code given from the previous assignment. 

main.cpp
#include "snake.h"
#include "map.h"
#include <cstdlib>
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <curses.h>
#include <vector>
#include <time.h>
#include <unistd.h>
using namespace std;

const char *victory=
" \n \n \n \n \n Yer a wizard! You've won! \n \n \n "
"            _            _.,----,\n "
" __   _.-._/ '_.        -  ,._ \\) \n "
"|  '-)_    '_.  \\      /   < _   )/   \n"
"//_    '_.   \\  '_, ___(c-(6)=(6) \n"
"   ''.    '.  '.  _,    >\\   ''  )\n"
"     |._.   '___'  (   ( ''/' =  '/ \n"
"        |..__    ,'-.')  '-  '__'  \n"
"  ''''''     /'._|    Y/     _/' \\ \n"
"    '''''\\  |   |   _| _.'._    \\  \n"
"  ''''''   \\|   \\  \\/     '._  \\ \n"
"   .-,-,_ ___|    '.  \\       \\  |  \n"
"  //               \\, \\______|   L \n "
" |                  \\_________\\ \\ \n" 
" '__'-'_.\\__/  |  .'/          \\ \\ \n"
"               [  (              \\ \\\n"
"               )  (               \\/\n"
"               \\_) \n \n \n \n \n";

const char *startingPage =
" \n"
"\n"
"\n"  
"    The king of the serpents... \n"
  " \n"
  "     |~~~~~~|     /\\      ~~~~~  | |     |  ~~~~~  |  / \n"
  "     |      |    /  \\     |        |        |      | /    \n "
  "    |~~~~~~|   /~~~~\\    |~~~~| | |     |  |~~~~| |<        \n"
  "     |      |  /      \\        | | |     |       | | \\   \n"
  "     |~~~~~~| /        \\  ~~~~~| | |~~~~ |  ~~~~~| |  \\ \n"
"\n"
"\n"
"\n"
  "                     Press 1 to play \n"
  "                Press 2 to get instructions \n"
  "              Or Press Any Other Key to Quit \n";

const char *loser=
"\n \n \n \n \n \n \n \n        Your loserness is an unforgivable curse... \n \n \n \n"
"       /\\   \\    /   /\\   ~~\\    /\\        |  /   /\\   ~~\\   /\\  \\    /  ~~\\   /\\ \n"
"      /--\\   \\  /   /--\\  |  |  /--\\       |<    /--\\  |  | /--\\  \\  /   |_|  /--\\ \n"
"     /    \\   \\/   /    \\ |_ / /    \\      |  \\ /    \\ |_/ /    \\  \\/    | \\ /    \\ \n"
"\n \n \n                                                    You're dead.  \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n";


const char *instructions=
"\n \n \n \n     So, you think you have what it takes to be a basilisk? We'll see... \n"
"     Voldemort has entered Hogwarts, and has trapped different muggle borns \n"
"     in the chamber of secrets. You must go around and petrify each of them\n"
"     in order to help Voldemort (and not be killed by him)\n \n \n"
"     To play, use the arrow keys to move up (or k), down (or j),\n"
"     left (or h), and right (or l)                                   \n"
"     But be quick, or you will start moving diagonally automatically \n"
"     Making it harder to petrify all of them. You can hold down the \n"
"     arrow keys to move faster. Also, don't try to go outside of the\n"
"     or you trip over yourself or you will be seen and killed. Good luck.\n \n \n"
"     Press 1 to continue, or any other key if you are too afraid \n";

int main(int argc, char *argv[]){
  srand(time(NULL));
  initscr();
  noecho();
  keypad(stdscr, TRUE);
  curs_set(0);
  cbreak();
  printw( startingPage);
  int key;
  switch(key=getch()){
  case '2':
    clear();
    printw(instructions);
    switch(key=getch()){
    case '1':
      break;
    default:
      endwin();
      return 1;
    }
  case '1':
    map_t m;
    new_map(&m);
    refresh();
    while((m.s->alive)==1){
      printMap(&m);
      do_move(&m);
      refresh();
      if(m.numTraps==0){
	endwin();
	delete_map(&m);
        printf("%s", victory);
	return 1;
      }
    }
    endwin();
    delete_map(&m);
    printf("%s",loser);   
  }
  return 1;

}



map.cpp
#include "snake.h"
#include "map.h"
#include <curses.h>
#include <stdlib.h>
#include <iostream>
#include <cstdlib>
#include <vector>
#include <stdio.h>
#include <unistd.h>
#include <err.h>
#include <time.h>

using namespace std;
//This code is given from my professor
typedef struct io_message{
char msg[71];
  struct io_message *next;
} io_message_t;
static io_message_t *io_head, *io_tail;

void io_queue_message(const char *format, ...)
{
  io_message_t *tmp;
  va_list ap;
  if (!(tmp = (io_message_t *) malloc(sizeof (*tmp)))) {
    perror("malloc");
    exit(1);
  }
  tmp->next = NULL;
  va_start(ap, format);
  vsnprintf(tmp->msg, sizeof (tmp->msg), format, ap);
  va_end(ap);
  if (!io_head) {
    io_head = io_tail = tmp;
  } else {
    io_tail->next = tmp;
    io_tail = tmp;
  }
}

static void io_print_message_queue(uint32_t y, uint32_t x)
{
  while (io_head) {
    io_tail = io_head;
    //attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(y, x, "%-80s", io_head->msg);
    //attroff(COLOR_PAIR(COLOR_CYAN));
    io_head = io_head->next;
    free(io_tail);
  }
  io_tail = NULL;
}

//Starting my own code
void new_map(map_t *m){
  int i,k, y, x;
  snake *sn= new snake();
  y=(rand()%46)+1;
  x=(rand()%46)+1;
  snake_set_alive(sn, 1);
  snake_set_x(sn,x);
  snake_set_y(sn,y);
  m->s=sn;
  m->numEggs=0;
  m->numTraps=20;
  for(i=0; i<50; i++){
    for(k=0; k<50; k++){
      m->EG[i][k]=0;
      m->traps[i][k]=0;
     }
   }
  for(i=0; i<m->numTraps; i++){
    y=rand()%46 +1;
    x=rand()%46 +1;
    while(m->traps[y][x]==1 ||( m->s->xposition==x && m->s->yposition==y)){
      y=rand()%46 +1;
      x=rand()%46 +1;
    }
    m->traps[y][x]=1;
  }
  
}

void delete_map(map_t *m){
  delete(m->s);
  int i;
  /*for(i=0; i<m->numEggs; i++){
    
    free(m->Eggs[i]);
    }*/
}

void do_move(map_t *m){
  move_eggs(m);
  fflush(stdout);
  int i;
  fd_set readfs;
  struct timeval tv;

    FD_ZERO(&readfs);
    FD_SET(0, &readfs);
    tv.tv_sec=0;
    tv.tv_usec=300000;
    switch(select(1, &readfs,NULL, NULL, &tv)){
    case -1:
    case 0:
      if(m->s->yOldPosition > m->s->yposition){
	snake_set_oldy(m->s, m->s->yposition);
	snake_set_y(m->s, (m->s->yposition)-1);
      }
      else{
	snake_set_oldy(m->s, m->s->yposition);
	  snake_set_y(m->s, (m->s->yposition)+1);
      }
      if(m->s->xOldPosition > m->s->xposition){
	  snake_set_oldx(m->s, m->s->xposition);
	  snake_set_x(m->s, (m->s->xposition)-1);
      }
      else{
	snake_set_oldx(m->s, m->s->xposition);
	snake_set_x(m->s, (m->s->xposition)+1);
      }
      break;
    case 1:
      switch(i=getch()){
      case KEY_LEFT:
      case 'h':
	snake_set_oldx(m->s, m->s->xposition);
	snake_set_x(m->s, (m->s->xposition)-1);
	break;
      case KEY_DOWN:
      case 'j':
	snake_set_oldy(m->s, m->s->yposition);
	snake_set_y(m->s, (m->s->yposition)+1);	
	break;
      case KEY_RIGHT:
      case 'l':
	snake_set_oldx(m->s, m->s->xposition);
	snake_set_x(m->s, (m->s->xposition)+1);	
	break;
      case KEY_UP:
      case 'k':
	snake_set_oldy(m->s, m->s->yposition);
	snake_set_y(m->s, (m->s->yposition)-1);	
	break;
      }
      break;
    }
    if(m->s->xposition==0 || m->s->xposition==48
       || m->s->yposition==0 || m->s->yposition==48){
      snake_set_alive(m->s, 0);
    }
    for(i=0; i<(m->numEggs); i++){
      if(m->Eggs[i].xposition==m->s->xposition
	 && m->Eggs[i].yposition==m->s->yposition){
	snake_set_alive(m->s, 0);
      }
    }
}
const char *mudbloods[]={
  "Hermione Granger",
  "Lily Potter",
  "Colin Creevey",
  "Mary Cattermole",
  "Moaning Myrtle",
  "Ted Tonks",
  "Justin Finch-Fletchley",
  "Mary Cattermole",
  "Dennis Creevey",
  "Dirk Cresswell",
  "Kendra Dumbledore",
  "Kevin Entwhistle",
  "Johannes Jonker",
  "Nobby Leach",
  "Garrick Ollivander's mother",
  "Tertius",
  "Donaghan Tremlett",
  "Wandless"
};
void move_eggs(map_t *m){
  int i,k;
  if(m->traps[m->s->yposition][m->s->xposition]){
    k=rand()%18;
    io_queue_message("You petrified %s.", mudbloods[k]);
   
    egg *e=new egg();
    e->xposition=m->s->xposition;
    e->yposition=m->s->yposition;
    if(m->numEggs==0){
      (m->Eggs).insert(m->Eggs.begin(), *e);
    }
    else{
     std::vector<egg>:: iterator it= m->Eggs.begin();
     (m->Eggs).insert(it+1,*e);
    }
    m->traps[m->s->yposition][m->s->xposition]=0;
    m->numTraps--;
    m->EG[e->yposition][e->xposition]=1;
    m->numEggs++;
  }
  else{
   
    for(i=0; i<m->numEggs; i++){
      if(i==0){
	m->EG[m->Eggs[i].yposition][m->Eggs[i].xposition]=0;
	(m->Eggs[i]).yposition=(m->Eggs[i+1]).yposition;
	(m->Eggs[i]).xposition=(m->Eggs[i+1]).xposition;
	m->EG[m->Eggs[i].yposition][m->Eggs[i].xposition]=1;
      }
      if(i==(m->numEggs -1)){
	m->Eggs[i].yposition=m->s->yposition;
	m->Eggs[i].xposition=m->s->xposition;
	m->EG[m->Eggs[i].yposition][m->Eggs[i].xposition]=1;
      }
      else{
	(m->Eggs[i]).yposition=(m->Eggs[i+1]).yposition;
	(m->Eggs[i]).xposition=(m->Eggs[i+1]).xposition;
	m->EG[m->Eggs[i].yposition][m->Eggs[i].xposition]=1;
      }
    }
  }
  
}

 void printMap(map_t *m){
   if(m->numEggs==0){
   clear();
   }
   init_pair(2, COLOR_GREEN, COLOR_BLACK);
   init_pair(6, COLOR_CYAN, COLOR_BLACK);
   int i,k;

   start_color();
   for(i=0; i<50; i++){
     for(k=0; k<50; k++){
       io_print_message_queue(1,51);
       if(i==0 && k==9){
	 mvaddch(i,k, ' ');
       }
       else if(i==0 && k==10){
	 attron(COLOR_PAIR(6));
	 mvaddch(i,k, 'C');
       }
       else if(i==0 && k==11){
	 mvaddch(i,k, 'h');
       }
       else if(i==0 && k==12){
	 mvaddch(i,k, 'a');
       }
       else if(i==0 && k==13){
	   mvaddch(i,k, 'm');
       }
       else if(i==0 && k==14){
	 mvaddch(i,k, 'b');
       }
       else if(i==0 && k==15){
	 mvaddch(i,k, 'e');
       }
       else if(i==0 && k==16){
	 mvaddch(i,k, 'r');
       }
       else if(i==0 && k==17){
	 mvaddch(i,k, ' ');
       }
       else if(i==0 && k==18){
	 mvaddch(i,k, 'o');
       }
       else if(i==0 && k==19){
	 mvaddch(i,k, 'f');
       }
       else if(i==0 && k==20){
	 mvaddch(i,k, ' ');
       }
       else if(i==0 && k==21){
	 mvaddch(i,k, 'S');
       }
       else if(i==0 && k==22){
	 mvaddch(i,k, 'e');
       }
       else if(i==0 && k==23){
	 mvaddch(i,k, 'c');
       }
       else if(i==0 && k==24){
	 mvaddch(i,k, 'r');
       }
       else if(i==0 && k==25){
	 mvaddch(i,k, 'e');
       }
       else if(i==0 && k==26){
	 mvaddch(i,k, 't');
       }
       else if(i==0 && k==27){
	 mvaddch(i,k, 's');
       }
       else if(i==0 && k==28){
	 mvaddch(i,k, ' ');
	 attroff(COLOR_PAIR(6));
       }
	      
       else if((i==0 && k!=49) || (k==0 && i!=49) || (i==48 && k!=49) || (k==48 && i!=49)){ 
	 mvaddch(i,k, 'X');
       }
       else if(m->s->yposition==i && m->s->xposition==k){
	 //green
	 attron(COLOR_PAIR(2));
	 mvaddch(i,k, 'S');
	 attroff(COLOR_PAIR(2));
       }
       
       else if(m->traps[i][k]==1){
	 mvaddch(i,k,'m');
       }
       else if(m->EG[i][k]==1){
	 attron(COLOR_PAIR(2));
	   mvaddch(i,k, '0');
	   attroff(COLOR_PAIR(2));
       }
       else{
	 mvaddch(i,k, ' ');
       }
       
     }
   }
 }
